phantom_token.c

```diff
@@ -628,54 +628,68 @@ static ngx_int_t introspection_response_handler(ngx_http_request_t *request, voi
         return introspection_subrequest_status_code;
     }
 
-    // body parsing
     u_char *jwt_start = NULL;
+    ngx_str_t cache_data = ngx_null_string;
 
+#if (NGX_HTTP_CACHE)
     if (!request->cache || !request->cache->buf)
     {
+        // No cache; read JWT from response to sub-request
         jwt_start = request->header_end + sizeof("\r\n") - 1;
     }
 
     if (jwt_start == NULL && request->cache && request->cache->buf && request->cache->valid_sec > 0)
     {
-        ngx_read_file(&request->cache->file, request->cache->buf->pos, request->cache->length, 0);
+        // Try to read JWT from cache
 
-        jwt_start = request->cache->buf->start + request->cache->body_start;
+        cache_data.len = request->cache->length;
+        cache_data.data = ngx_pnalloc(request->pool, cache_data.len);
```

Comment:
Is it OK to use `ngx_pnalloc` instead of `ngx_pcalloc` here?

Reply:
I think so:

> ngx_pnalloc(pool, size) — Allocate unaligned memory from the specified pool. Mostly used for allocating strings.

https://nginx.org/en/docs/dev/development_guide.html#memory_management

Not sure why we'd need the memory to be aligned 

> ngx_pcalloc(pool, size) — Allocate aligned memory from the specified pool and fill it with zeroes.

I guess what we'd want is that it's zeroed out, but this doesn't matter, I don't think, because it's a string and we have the length.

Reply:
I was mostly concerned about zeroing it (just because we were doing that before).

Reply:
I'll change it to zero out the mem and to be consistent. Look for a commit before merging. Thanks for the inputs, @andrei-curity .

Reply:
after double checking, I decided to leave this even though the suggestion isn't a bad one. Reason is:

1. It's what's now been tested at a customer under high loads
2. There's quite a few uses `ngx_pnalloc` in addition to `ngx_pcalloc`, so it's consistent either way.
3. I'm not sure what the NGINX docs mean by "aligned" and don't wan to waste/fragment memory when working with strings -- which the docs point out is the general use of `ngx_pnalloc`.